pkg/controller/identity/iamrolepolicyattachment/controller.go

```diff
@@ -125,90 +123,61 @@ func (e *external) Observe(ctx context.Context, mgd resource.Managed) (managed.E
 		}, nil
 	}
 
+	current := cr.Spec.ForProvider.DeepCopy()
+	iam.LateInitializePolicy(&cr.Spec.ForProvider, attachedPolicyObject)
+	if !reflect.DeepEqual(current, &cr.Spec.ForProvider) {
+		if err := e.kube.Update(ctx, cr); err != nil {
+			return managed.ExternalObservation{}, errors.Wrap(err, errKubeUpdateFailed)
+		}
+	}
+
 	cr.SetConditions(runtimev1alpha1.Available())
 
-	cr.UpdateExternalStatus(*attachedPolicyObject)
+	cr.Status.AtProvider = iam.GenerateRolePolicyObservation(*attachedPolicyObject)
 
 	return managed.ExternalObservation{
-		ResourceExists: true,
+		ResourceExists:   true,
+		ResourceUpToDate: cr.Spec.ForProvider.PolicyARN == cr.Status.AtProvider.AttachedPolicyARN,
 	}, nil
 }
 
 func (e *external) Create(ctx context.Context, mgd resource.Managed) (managed.ExternalCreation, error) {
-	cr, ok := mgd.(*v1alpha3.IAMRolePolicyAttachment)
+	cr, ok := mgd.(*v1beta1.IAMRolePolicyAttachment)
 	if !ok {
 		return managed.ExternalCreation{}, errors.New(errUnexpectedObject)
 	}
 
 	cr.SetConditions(runtimev1alpha1.Creating())
 
-	req := e.client.AttachRolePolicyRequest(&awsiam.AttachRolePolicyInput{
-		PolicyArn: aws.String(cr.Spec.PolicyARN),
-		RoleName:  aws.String(cr.Spec.RoleName),
-	})
+	req := e.client.AttachRolePolicyRequest(iam.GenerateAttachRolePolicyInput(&cr.Spec.ForProvider))
+	req.SetContext(ctx)
 
 	_, err := req.Send(ctx)
 
-	return managed.ExternalCreation{}, errors.Wrapf(err, errAttach, cr.Spec.PolicyARN, cr.Spec.RoleName)
+	return managed.ExternalCreation{}, errors.Wrapf(err, errAttach, cr.Spec.ForProvider.PolicyARN, cr.Spec.ForProvider.RoleName)
```

Comment:
You don't need to return anything other than the error(`err`) and the error string(`errAttach`) as managed reconciler properly fills the error and logs with necessary identifiers.

Reply:
@muvaf `errAttach` has format specifiers, and removing the values for those resulted in this for me (kubectl describe)
```
 Warning  CannotCreateExternalResource  6m59s  managed/iamrolepolicyattachment.identity.aws.crossplane.io  failed to attach the policy with arn %!!(MISSING)v(MISSING) to role %!!(MISSING)v(MISSING): NoSuchEntity: Policy arn:aws:iam::aws:policy/AleaForBusinessDeviceSetup does not exist or is not attachable.
```
The values are not filled and I am not sure how they will. Am I missing something?

Reply:
I believe @muvaf just means that you can remove the formatted string and return `errors.Wrap(err, errAttach)` :)

Reply:
```suggestion
	return managed.ExternalCreation{}, errors.Wrap(err, errAttach)
```

I'm suggesting this because this. We don't need to repeat what's on the CR as long as the error string is specific enough.